Ted vision

asherbolotski opened the pull request.

mcadamswa committed:
adding camera

mcadamswa committed:
Update CameraSubsystem.java

asherbolotski committed:
Update build.gradle

asherbolotski committed:
ported relative to tag command to ted

asherbolotski committed:
added fmap

asherbolotski committed:
added get distance from tag command

mcadamswa committed:
.

mcadamswa committed:
Merge branch 'main' into ted_vision

mcadamswa committed:
working around a git issue revert this commit.

mcadamswa committed:
.

mcadamswa committed:
Merge branch 'main' into ted_vision

asherbolotski committed:
added auto yaw command

josieammer committed:
handle case where tag is not in view in ShooterSetAngleWithVisionCommand

mcadamswa committed:
Update CameraSubsystem.java

mcadamswa committed:
added peicewise

asherbolotski committed:
made so distance command can look at two tags

mcadamswa committed:
Create Pipeline_Name (1).vpr

asherbolotski committed:
Merge branch 'main' into ted_vision

asherbolotski committed:
Update CameraSubsystem.java

asherbolotski committed:
.

asherbolotski committed:
Update RobotContainer.java

asherbolotski committed:
Update RobotContainer.java

asherbolotski committed:
.

asherbolotski committed:
Update ShooterSetAngleWithVisionCommand.java

asherbolotski committed:
Update ShooterSetAngleWithVisionCommand.java

asherbolotski committed:
Update CameraSubsystem.java

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:
```diff
@@ -27,9 +27,11 @@
 import frc.robot.common.EulerAngle;
 import frc.robot.common.VectorUtils;
 import frc.robot.control.SwerveDriveMode;
+import frc.robot.control.SubsystemCollection;
```
Seems like there are conflicts with main ... I will do merge into your branch.

mmcadamsms replied:
I merged

mmcadamsms committed:
Merge branch 'main' into ted_vision

mmcadamsms commented on ted/src/main/java/frc/robot/commands/AllignRelativeToTagCommand.java:
```diff
@@ -0,0 +1,129 @@
+  /**
+   * Constructor for command.
+   */
+  public AllignRelativeToTagCommand(DrivetrainSubsystem drivetrainSubsystem, CameraSubsystem camerasubsystem, Pose2d targetPoseInVisionSpace, double tagId) {
```
Need proper javadoc

asherbolotski replied:
👍

mmcadamsms commented on ted/src/main/java/frc/robot/commands/AllignRelativeToTagCommand.java:
```diff
@@ -0,0 +1,129 @@
+/**
+ * Implements a command.
+ */
+public class AllignRelativeToTagCommand extends Command{
```
need better javadoc

asherbolotski replied:
done

mmcadamsms commented on ted/src/main/java/frc/robot/commands/AllignRelativeToTagCommand.java:
```diff
@@ -0,0 +1,129 @@
+      done = true;
+    }
+
+    drivetrainsubsystem.drive(new ChassisSpeeds(0, 0, 0));
```
So seemingly this command can be used only when the robot is still?

asherbolotski replied:
this just sets the chassis speeds to 0 on init so we aren't actively driving while we try to run the command, we can still call the command while the robot is moving

mmcadamsms commented on ted/src/main/java/frc/robot/commands/AllignRelativeToTagCommand.java:
```diff
@@ -0,0 +1,129 @@
+        rotVelocity = -1 * MotorUtils.clamp(rotVelocity, -velocityFactor, velocityFactor);
+      }
+
+      System.out.println("xVelocity: " + xVelocity);
```
This will be a fair bit of printing ... need to reduce spew

asherbolotski replied:
removed

mmcadamsms commented on ted/src/main/java/frc/robot/commands/AllignRelativeToTagCommand.java:
```diff
@@ -0,0 +1,129 @@
+      drivetrainsubsystem.drive(new ChassisSpeeds(xVelocity, yVelocity, rotVelocity));
+    }
+
+
```
extra line

asherbolotski replied:
removed

mmcadamsms commented on ted/src/main/java/frc/robot/commands/AllignRelativeToTagCommand.java:
```diff
@@ -0,0 +1,129 @@
+    //todo
+    // for the rotation, you need to wrap the angle around using angleModulus. angleModulus operates in radians, 
+    // so you will need to wrap the angle in radians and then convert to degrees.
+    done = Units.radiansToDegrees(Math.abs(MathUtil.angleModulus((targetPositionInRobotSpace.getRotation().getRadians() - startingPositionInRobotSpace.getRotation().getDegrees())))) <= rotationTolerance
```
shorter lines ... 120 character boundary is customary

asherbolotski replied:
removed

mmcadamsms commented on ted/src/main/java/frc/robot/commands/AllignRelativeToTagCommand.java:
```diff
@@ -0,0 +1,129 @@
+  @Override
+  public boolean isFinished(){
+    startingPositionInRobotSpace = drivetrainsubsystem.getRobotPosition();
+    //todo
```
TODO's are usually capitalized as these tend to stand out

asherbolotski replied:
removed

mmcadamsms commented on ted/src/main/java/frc/robot/commands/ShooterSetAngleWithVisionCommand.java:
```diff
@@ -0,0 +1,76 @@
+        slope = farSlope;
+        offset = farOffset;
+      }
+      desiredAngleDegrees = (slope*cameraSubsystem.getDistanceFromTag(7.0, 4.0).getDistanceMeters()) + offset;
```
seems like part of this line is repetitive ... can we use the stack variable from line 58 here?

asherbolotski replied:
fixed

mmcadamsms commented on ted/src/main/java/frc/robot/commands/ShooterSetAngleWithVisionCommand.java:
```diff
@@ -0,0 +1,76 @@
+
+  @Override
+  public void execute() {
+    DistanceMeasurement distanceMeasurement = cameraSubsystem.getDistanceFromTag(7.0, 4.0);
```
magic numbers

asherbolotski replied:
resolved

mmcadamsms commented on ted/src/main/java/frc/robot/commands/ShooterSetAngleWithVisionCommand.java:
```diff
@@ -0,0 +1,76 @@
+  private double desiredAngleDegrees;
+  private ShooterAngleSubsystem shooterAngleSubsystem;
+  private CameraSubsystem cameraSubsystem;
+  private double farSlope = -4.65;
```
curious to know what is going on here with these slopes and offsets ... not understanding from the code

asherbolotski replied:
added explanation

mmcadamsms commented on ted/src/main/java/frc/robot/commands/UseVisionForPositionCommand.java:
```diff
@@ -0,0 +1,67 @@
+
+package frc.robot.commands;
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
```
update to Command

mmcadamsms commented on ted/src/main/java/frc/robot/commands/UseVisionForPositionCommand.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Implements a command
+ */
+public class UseVisionForPositionCommand extends CommandBase{
```
update to Command

asherbolotski replied:
updated

mmcadamsms commented on ted/src/main/java/frc/robot/commands/UseVisionForPositionCommand.java:
```diff
@@ -0,0 +1,67 @@
+// ************************************************************
```
Seems like this is not in use in the codebase in this PR.  When will it be used?  Should it be added then?

asherbolotski replied:
this command allows us to update our odometry with vision, so if we wanted to do this in auto or for yaw calculations it would be useful.

josieammer replied:
this doesn't yet work.  you should remove this file to a new branch for future testing.

asherbolotski replied:
removed

mmcadamsms commented on ted/src/main/java/frc/robot/commands/AllignRelativeToTagCommand.java:
```diff
@@ -0,0 +1,129 @@
+// ************************************************************
```
Seems like this is not in use in the codebase in this PR.  When will it be used?  Should it be added then?

asherbolotski replied:
We might use this for autonomous, allowing us to have more accurate autos because if we bump something our the robot would still know where it is because of vision

josieammer replied:
this command doesn't yet work. Pull this file out of the PR and put it on a separate branch for future testing

asherbolotski replied:
removed

mmcadamsms reviewed (commented):
first set of comments up through commands

mmcadamsms commented on ted/src/main/java/frc/robot/common/DriveTrainConfig.java:
```diff
@@ -0,0 +1,109 @@
+/**
+ * Forms a type to hold drivetrian config
+ */
+public class DrivetrainConfig {
```
So casing on the Java file name is DriveTrainConfig and the class name is DrivetrainConfig.  One of these should be changed.

asherbolotski replied:
fixed

mmcadamsms commented on ted/src/main/java/frc/robot/common/DriveTrainConfig.java:
```diff
@@ -0,0 +1,109 @@
+// Bishop Blanchet Robotics
+// Home of the Cybears
+// FRC - Crescendo 2024
+// File: DrivetrainConfig.java
```
So this class essentially already exists.  See: frc.robot.common.DrivetrainConfig

mmcadamsms replied:
so seems like its the casing that is messing things up ... need to figure out casing issue here

asherbolotski replied:
adressed

josieammer replied:
the file should be renamed DrivetrainConfig.java.

mmcadamsms commented on ted/src/main/java/frc/robot/common/DriveTrainConfig.java:
```diff
@@ -0,0 +1,109 @@
+    private double backLeftModuleSteerOffset;
+    private double backRightModuleSteerOffset;
+
+    public DrivetrainConfig(
```
Need proper java doc comments

mmcadamsms commented on ted/src/main/java/frc/robot/common/DriveTrainConfig.java:
```diff
@@ -0,0 +1,109 @@
+
+/**
+ * Forms a type to hold drivetrian config
+ */
```
need correct javadoc comments

mmcadamsms commented on ted/src/main/java/frc/robot/common/DriveTrainConfig.java:
```diff
@@ -0,0 +1,109 @@
+        this.trackwidthMeters = trackwidthMeters;
+    }
+
+    public void setWheelbaseMeters(double wheelbaseMeters) {
```
javadoc comments

mmcadamsms commented on ted/src/main/java/frc/robot/control/ManualInputInterfaces.java:
```diff
@@ -19,21 +19,7 @@
+import frc.robot.commands.*;
 
 public class ManualInputInterfaces {
 
```
I had expected to see the new command hooked in here around about line #362 (old variant with individually specified import commands maybe 15 lines further down).

asherbolotski replied:
added to button

mmcadamsms commented on ted/src/main/java/frc/robot/control/SubsystemCollection.java:
```diff
@@ -26,6 +27,7 @@ public class SubsystemCollection
     // declaring and init subsystems  
     private ClimberSubsystem climberSubsystem = null;
     private DrivetrainSubsystem driveTrainSubsystem = null;
+    private CameraSubsystem cameraSubsystem = null;
```
Reorder alphabetically

asherbolotski replied:
fixed

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/CameraSubsystem.java:
```diff
@@ -3,51 +3,118 @@
+  private final int botPositionYIndex = 2;
+  private final int botRotationIndex = 5;
+  private final int noTagInSightId = -1;
+  NetworkTable table = NetworkTableInstance.getDefault().getTable("limelight");
```
Why do we need a public NetworkTable member?  If you don't need it exposed, please make it private

asherbolotski replied:
made private

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/CameraSubsystem.java:
```diff
@@ -3,51 +3,118 @@
+  NetworkTable table = NetworkTableInstance.getDefault().getTable("limelight");
+  /**
+   * a constructor for the camera subsystem class
+   * @param subsystems - the subsystem collection
```
this is default ctor, no need for a @param here

asherbolotski replied:
removed

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/CameraSubsystem.java:
```diff
@@ -3,51 +3,118 @@
-    return false;
+  public Pose2d getVisionBotPoseInTargetSpace(){
+    double tagId = table.getEntry("tid").getDouble(0);
+    double[] botpose = table.getEntry("botpose_targetspace").getDoubleArray(new double[defaultDoubleArraySize]);
```
While seemingly not required, using 'this.' keyword when referencing member variables helps the reader understand that the variable here is not stack, but instead its heap (member).

mmcadamsms replied:
as in ...

double[] botpose = this.table.getEntry("botpose_targetspace").getDoubleArray(new double[this.defaultDoubleArraySize]);

asherbolotski replied:
added this's everywhere

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/CameraSubsystem.java:
```diff
@@ -3,51 +3,118 @@
-   *
-   * @return value of some boolean subsystem state, such as a digital sensor.
+   * a method that returns a vision measurement. 
+   * pose portion of the vision measurement is null if there is no valid measurement. 
```
update for proper javadoc comments

asherbolotski replied:
updated

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/CameraSubsystem.java:
```diff
@@ -3,51 +3,118 @@
+  public double getTagId(){
+    return table.getEntry("tid").getDouble(0);
+  }
+
```
proper javadoc comments

asherbolotski replied:
added

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/CameraSubsystem.java:
```diff
@@ -3,51 +3,118 @@
+      return new VisionMeasurement(realRobotPosition, timestamp);
+    }
+  }
+
```
proper javadoc comments

asherbolotski replied:
added

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:
```diff
@@ -145,7 +157,9 @@ public class DrivetrainSubsystem extends SubsystemBase {
    */
-  public DrivetrainSubsystem() {
+  public DrivetrainSubsystem(SubsystemCollection subsystems) {
+    cameraSubsystem = subsystems.getCameraSubsystem();
```
Need to couch this in an if block properly.  Subsystems come and go around here.  We need to make sure it is present before assignment.

asherbolotski replied:
added a installed hardware check

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:
```diff
@@ -43,17 +45,27 @@
+import edu.wpi.first.math.MatBuilder;
 
 public class DrivetrainSubsystem extends SubsystemBase {
+  CameraSubsystem cameraSubsystem;
```
This should not be public scope.

asherbolotski replied:
made private

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:
```diff
@@ -258,6 +272,10 @@ public SwerveDriveKinematics getSwerveKinematics() {
     return swerveKinematics;
   }
 
+  public void setUseVision(boolean b){
```
proper comments and not super sure what b means to the eventual caller of this api, names should be more self explanatory

asherbolotski replied:
updated

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:
```diff
@@ -634,6 +654,20 @@ public void zeroRobotPosition()
+  private void addVisionMeasurement(VisionMeasurement visionMeasurement){
+    // The wpilib matrix constructor requires sizes specified as Nat types. 
+    // trying to update the matrix dynamicaly leads to issues
+    Matrix<N3,N1> visionStdDev = (new MatBuilder<N3,N1>(Nat.N3(), Nat.N1())).fill(new double[]{0.9, 0.9, 0.9});
```
MatBuilder is deprecated in 2024.  Find the proper replacement.

josieammer replied:
replacement is here: https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/math/MatBuilder.html#fill(edu.wpi.first.math.Nat,edu.wpi.first.math.Nat,double...)

asherbolotski replied:
fixed

asherbolotski replied:
?

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:
```diff
@@ -634,6 +654,20 @@ public void zeroRobotPosition()
+    // trying to update the matrix dynamicaly leads to issues
+    Matrix<N3,N1> visionStdDev = (new MatBuilder<N3,N1>(Nat.N3(), Nat.N1())).fill(new double[]{0.9, 0.9, 0.9});
+    // for now ignore all vision measurements that are null or contained robot position is null
+    if (visionMeasurement != null && visionMeasurement.getRobotPosition() != null && useVision){
```
Is the cost of visionMeasurement.getRobotPosition() low?  Is the memory cost of what it returns high?  Is it possible the state changes between successive calls into visionMeasurement.getRobotPosition()? Your code assumes the compute cost of the method is very low, the memory cost might be high-ish and that successive recompute of getRobotPosition will return the same result.

I'd bet that the above are less likely to be true (but then again, I'm not totally sure either).  Usually if I'm using a variable 2x from a method in quick succession, I decide to pay for the memory (it needs to be on the stack at least momentarily anyway, why not keep it on for just a while longer), save the compute and therefore gain consistency between usages. 

In essence, I'd imagine that I'd use code that looks like this instead:

    // for now ignore all vision measurements that are null or contained robot position is null
    if (visionMeasurement != null && useVision){
      Pose2d visionComputedMeasurement = visionMeasurement.getRobotPosition();
      if(visionComputedMeasurement != null) {
        swervePoseEstimator.addVisionMeasurement(visionMeasurement.getRobotPosition(), visionMeasurement.getTimestamp(), visionStdDev);
      }
    }

PS - My references to above stack/heap might all turn out to be substantially heap-based in java-land, but the concept remains, more stuff we put on the heap means the GC needs to do more work, more often.

asherbolotski replied:
resolved

mmcadamsms commented on ted/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:
```diff
@@ -485,6 +503,8 @@ public void periodic() {
     // refresh the position of the robot
     this.refreshRobotPosition();
+    // update robot position with vision
+    this.addVisionMeasurement(cameraSubsystem.getVisionBotPose());
```
Need to protect this line against null reference exception when vision is not in use.

asherbolotski replied:
already in place with the "useVision" boolean.

asherbolotski replied:
added a installed hardware check

josieammer committed:
replace ShooterSetAngleWithVisionCommand with two versions

one version is for continuous operation (e.g. attached to a UI button)
the second version is one-shot operation (e.g. for using in autonomous)

asherbolotski committed:
first batch of PR comments

asherbolotski committed:
first set of pr addresses

mmcadamsms committed:
Merge branch 'main' into ted_vision

mcadamswa committed:
Merge branch 'main' into ted_vision

asherbolotski committed:
Delete AllignRelativeToTagCommand.java

josieammer commented on ted/src/main/java/frc/robot/common/ShooterAngleHelpers.java:
```diff
@@ -0,0 +1,40 @@
+
+import frc.robot.control.Constants;
+
+public class ShooterAngleHelpers {
```
needs javadoc

asherbolotski replied:
done

josieammer commented on ted/src/main/java/frc/robot/common/ShooterAngleHelpers.java:
```diff
@@ -0,0 +1,40 @@
+    private static double closeOffset = 79.1;
+    private static double nearFarBreakpoint = 2.0;
+
+    public static double shooterAngleFromDistance(double distance){
```
needs javadoc

asherbolotski replied:
done

josieammer commented on ted/src/main/java/frc/robot/subsystems/CameraSubsystem.java:
```diff
@@ -3,51 +3,128 @@
 // FRC - Crescendo - 2024
 // File: CameraSubsystem.java
-// Intent: Forms a stub for the prelminary named subsystem above.
+// Intent: Forms the prelminary code for camera train subsystem.
```
fix this comment

asherbolotski replied:
done

josieammer reviewed (changes requested):
just a few more fixes needed

asherbolotski committed:
pr comments

josieammer committed:
renaming this file to workaround Git casing issue

josieammer reviewed (approved):
LGTM
